sys/net/gnrc/link_layer/lwmac/rx_state_machine.c

```diff
@@ -0,0 +1,393 @@
+/*
+ * Copyright (C) 2015 Daniel Krebs
+ *               2016 INRIA
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_lwmac
+ * @{
+ *
+ * @file
+ * @brief       Implementation of RX state machine of LWMAC protocol
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @author      Shuguo Zhuo  <shuguo.zhuo@inria.fr>
+ * @}
+ */
+
+#include "net/gnrc.h"
+#include "net/gnrc/lwmac/lwmac.h"
+#include "net/gnrc/mac/internal.h"
+#include "net/gnrc/lwmac/timeout.h"
+#include "include/rx_state_machine.h"
+#include "include/lwmac_internal.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define LOG_LEVEL LOG_WARNING
+#include "log.h"
+
+#undef LOG_ERROR
+#undef LOG_WARNING
+#undef LOG_INFO
+#undef LOG_DEBUG
+
+#define LOG_ERROR(...) LOG(LOG_ERROR, "ERROR: [lwmac-rx] " __VA_ARGS__)
+#define LOG_WARNING(...) LOG(LOG_WARNING, "WARNING: [lwmac-rx] " __VA_ARGS__)
+#define LOG_INFO(...) LOG(LOG_INFO, "[lwmac-rx] " __VA_ARGS__)
+#define LOG_DEBUG(...) LOG(LOG_DEBUG, "[lwmac-rx] " __VA_ARGS__)
+
+/* Break out of switch and mark the need for rescheduling */
+#define GOTO_RX_STATE(rx_state, do_resched) gnrc_netdev->rx.state = rx_state; \
+                                            reschedule = do_resched; \
+                                            break
+
+void lwmac_rx_start(gnrc_netdev_t *gnrc_netdev)
+{
+    if (gnrc_netdev == NULL) {
+        return;
+    }
+
+    /* RX address should have been reset, probably not stopped then */
+    assert(gnrc_netdev->rx.l2_addr.len == 0);
+
+    /* Don't attempt to send a WA if channel is busy to get timings right */
+    netopt_enable_t csma_disable = NETOPT_DISABLE;
+    gnrc_netdev->dev->driver->set(gnrc_netdev->dev, NETOPT_CSMA, &csma_disable,
+                                   sizeof(csma_disable));
+
+    gnrc_netdev->rx.state = RX_STATE_INIT;
+}
+
+void lwmac_rx_stop(gnrc_netdev_t *gnrc_netdev)
+{
+    if (!gnrc_netdev) {
+        return;
+    }
+
+    lwmac_clear_timeout(gnrc_netdev, TIMEOUT_DATA);
+    gnrc_netdev->rx.state = RX_STATE_STOPPED;
+    gnrc_netdev->rx.l2_addr.len = 0;
+}
+
+/* Returns whether rescheduling is needed or not */
+static bool _lwmac_rx_update(gnrc_netdev_t *gnrc_netdev)
+{
+    bool reschedule = false;
```

Comment:
@miri64 Trying to find a way to divide the these large functions efficiently with small revisions, but failed. 

The whole Lw-MAC adopts such a way to run all of its main manipulations (all over Rx state machines Tx state-machine and Lw-MAC's main body state-machine). I think/guess, probably, Daniel originally wants to introduce as low function-calling delays as possible, especially at MAC layer that is based on exact timing and requires fast reactions, such that he decided not to divide those large functions.

On the other hand, the current structure is, at least, already stable, seems to me as shown by my tests on LW-MAC. To this end, I will leave the structure as it is now and quit the division attempt on those big functions (i.e., the main bodies of Lw-MAC), which may introduce another round of massive tests and verifications.